File: src/vct/col/ast/ASTList.java
package vct.col.ast;

import java.util.ArrayList;
import java.util.Iterator;

public class ASTList implements ASTSequence<ASTList> {

  private ArrayList<ASTNode> list=new ArrayList<ASTNode>();
  
  @Override
  public Iterator<ASTNode> iterator() {
    return list.iterator();
  }

  @Override
  public ASTList add(ASTNode item) {
    if (item!=null) list.add(item);
    return this;
  }

  @Override
  public int size() {
    return list.size();
  }

  @Override
  public ASTNode get(int i) {
    return list.get(i);
  }

}

File: vercors/src/main/java/vct/col/ast/generic/ASTList.scala
package vct.col.ast.generic

import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer

class ASTList extends ASTSequence[ASTList] {
  private[this] val list = new ArrayBuffer[ASTNode]()
  
  def add(item:Option[ASTNode]) = item match {
    case Some(node) => list += node; this
    case None => this
  }
  
  override def add(item:ASTNode) = add(Option(item))
  override def iterator = list.toIterator.asJava
  override def get(i:Int) = list.apply(i)
  override def size = list.size
}
